### PR TITLE
spi: cleanup designware driver to be more devicetree oriented

### PR DESCRIPTION
**Pull request by galak**

### CODE REVIEW

**Review by pabigot — COMMENTED** [review]
Removing the application `CONFIG_SPI_x_OP_MODES` seems questionable, as external users might be using that to select peripheral (slave) vs controller (master) mode, and the driver does appear to support both.  I don't know if that's a use case that needs to be handled.

This appears to be the only driver that uses `CONFIG_SPI_#_OP_MODES`.  If we're going to remove the use, perhaps the Kconfigs should also be removed.

I'm not convinced mode selection belongs in devicetree.  It could be that inferring the mode from spi_configure() is the better path.

**Comment by galak** [comment]
> Removing the application `CONFIG_SPI_x_OP_MODES` seems questionable, as external users might be using that to select peripheral (slave) vs controller (master) mode, and the driver does appear to support both. I don't know if that's a use case that needs to be handled.
> 
> This appears to be the only driver that uses `CONFIG_SPI_#_OP_MODES`. If we're going to remove the use, perhaps the Kconfigs should also be removed.
> 
> I'm not convinced mode selection belongs in devicetree. It could be that inferring the mode from spi_configure() is the better path.

So the only use of CONFIG_SPI_x_OP_MODES in the driver is to specify if the hardware itself supports the mode.  That seems exactly what devicetree would describe.

Hopefully someone from arc or intel can chime in about what their specific hardware is doing here.

**Comment by pabigot** [comment]
> So the only use of CONFIG_SPI_x_OP_MODES in the driver is to specify if the hardware itself supports the mode. That seems exactly what devicetree would describe.

I'm not convinced that's the case.  In at least one similar driver I was reviewing recently the issue was that the hardware supports both, but only one at a time, so you have to pick at the time of initialization.

But if this one  is like Nordic where there are really two different peripherals that happen to share a driver for convenience then yes, it'd be devicetree to select which is which.

**Comment by galak** [comment]
> I'm not convinced that's the case. In at least one similar driver I was reviewing recently the issue was that the hardware supports both, but only one at a time, so you have to pick at the time of initialization.
> 
> But if this one is like Nordic where there are really two different peripherals that happen to share a driver for convenience then yes, it'd be devicetree to select which is which.

We'll the value is constant per instance, so feels like a hardware config of how the IP might have been built, vs a runtime behavior:

https://github.com/zephyrproject-rtos/zephyr/blob/b5a3fe14713a9f9662476337a71b29ba569a45a0/drivers/spi/spi_dw.c#L565

**Review by dcpleung — APPROVED** [review]

**Comment by galak** [comment]
@ruuddw / @abrodkin ping.

File: drivers/spi/spi_dw.c [hunk 1]
```diff
@@ -521,9 +521,6 @@ int spi_dw_init(const struct device *dev)
 	const struct spi_dw_config *info = dev->config;
 	struct spi_dw_data *spi = dev->data;
 
-	clock_config(dev);
```
Comment (ruuddw) [inline drivers/spi/spi_dw.c:524]: Can the #include <drivers/clock_control.h> be removed, without the clock control?

**Comment by ruuddw** [comment]
The DesignWare SPI controller can be configured either as master or slave at hardware design time -> no need for SW configurability. A DTS option would be nice, but I'm ok with current hardcoded version. Several ARC development boards have multiple SPI instances, e.g. one master and one slave. But, slave mode is currently not actively supported for these in Zephyr.

**Comment by galak** [comment]
> The DesignWare SPI controller can be configured either as master or slave at hardware design time -> no need for SW configurability. A DTS option would be nice, but I'm ok with current hardcoded version. Several ARC development boards have multiple SPI instances, e.g. one master and one slave. But, slave mode is currently not actively supported for these in Zephyr.

Sounds good, seems like a dts property than would make sense in the future to configure the setting per instance.

File: drivers/spi/spi_dw.c (hunk 1 above)
Comment (galak) [reply drivers/spi/spi_dw.c:524]: removed the include.

**Review by ruuddw — APPROVED** [review]
Thanks Kumar, indeed dts would be the right place to add back a slave configuration option. Let me check if Synopsys wants to do that to add slave support for some developement boards. That would be beyond the scope of this PR.

**Comment by abrodkin** [comment]
@galak unfortunately I don't have much to say here as of today we don't have any harness for SPI testing on our boards, even though on majority of our boards we do have DW SPI controllers (which for the record is DesignWare SSI "Synchronous Serial Interface"). We do plan to implement it in foreseeable future, but for now I guess we don't even build SPI drivers for ARC boards, sorry.

**Pull request merged**